<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.8.1" />
<title>dadi API documentation</title>
<meta name="description" content="For examples of dadi&#39;s usage, see the examples directory in the source
distribution …" />
<link href='https://cdnjs.cloudflare.com/ajax/libs/normalize/8.0.0/normalize.min.css' rel='stylesheet'>
<link href='https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/8.0.0/sanitize.min.css' rel='stylesheet'>
<link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" rel="stylesheet">
<style>.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Package <code>dadi</code></h1>
</header>
<section id="section-intro">
<p>For examples of dadi's usage, see the examples directory in the source
distribution.</p>
<p>Documentation of all methods can be found in doc/api/index.html of the source
distribution.</p>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">&#34;&#34;&#34;
For examples of dadi&#39;s usage, see the examples directory in the source
distribution.

Documentation of all methods can be found in doc/api/index.html of the source
distribution.
&#34;&#34;&#34;
import logging
logging.basicConfig()

from . import Demographics1D, Demographics2D, Inference, Integration
from . import Misc, Numerics, PhiManip, Spectrum_mod, tridiag, Godambe

# We do it this way so it&#39;s easier to reload.
Spectrum = Spectrum_mod.Spectrum

# Protect import of Plotting in case matplotlib not installed.
try:
    from . import Plotting
except ImportError:
    pass

# When doing arithmetic with Spectrum objects (which are masked arrays), we
# often have masked values which generate annoying arithmetic warnings. Here
# we tell numpy to ignore such warnings. This puts greater onus on the user to
# check results, but for our use case I think it&#39;s the better default.
import numpy
numpy.seterr(all=&#39;ignore&#39;)

def cuda_enabled(toggle=None):
    &#34;&#34;&#34;
    Enable or disable cuda execution
    &#34;&#34;&#34;
    if toggle is None:
        return Integration.cuda_enabled
    elif toggle == True:
        try:
            from . import cuda
            Integration.cuda_enabled = True
            return True
        except:
            print(&#34;Failed to import dadi.cuda&#34;)
            return False
    elif toggle == False:
        Integration.cuda_enabled = False
        return False
    else:
        raise ValueError(&#34;toggle must be True, False, or None&#34;)

__pdoc__ = {&#39;cuda&#39;:False}

def pts_to_RAM(pts, P):
    &#34;&#34;&#34;
    Approximate RAM usage for a given grid points and number of populations 

    pts: Grid points setting
    P: Number of populations
    &#34;&#34;&#34;
    return 8*4*pts**P / 1024**3

def RAM_to_pts(RAM, P):
    &#34;&#34;&#34;
    Approximate maximum grid points given the number of populations and available RAM

    pts: Grid points setting
    P: Number of populations
    &#34;&#34;&#34;
    return int((RAM*1024**3/(8*4))**(1./P))

def citation():
    &#34;&#34;&#34;
    Print citation information for dadi codebase.
    &#34;&#34;&#34;
    print(&#34;&#34;&#34;
If you find dadi useful in your research, please cite:
RN Gutenkunst, RD Hernandez, SH Williamson, CD Bustamante &#34;Inferring the joint demographic history of multiple populations from multidimensional SNP data&#34; PLoS Genetics 5:e1000695 (2009)

If you find the Godambe Information Matrix methods useful, please cite:
AJ Coffman, P Hsieh, S Gravel, RN Gutenkunst &#34;Computationally efficient composite likelihood statistics for demographic inference&#34; Molecular Biology and Evolution 33:591 (2016)

If you find the DFE inference methods useful, please cite:
BY Kim, CD Huber, KE Lohmueller &#34;Inference of the Distribution of Selection Coefficients for New Nonsynonymous Mutations Using Large Samples&#34; Genetics 206:345 (2017)

If you find the triallelic methods useful, please cite:
AP Ragsdale, AJ Coffman, P Hsieh, TJ Struck, RN Gutenkunst &#34;Triallelic population genomics for inferring correlated fitness effects of same site nonsynonymous mutations&#34; Genetics 203:513 (2016)

If you find the two-locus methods useful, please cite:
AP Ragsdale, RN Gutenkunst &#34;Inferring demographic history using two-locus statistics&#34; Genetics 206:1037 (2017)
&#34;&#34;&#34;)</code></pre>
</details>
</section>
<section>
<h2 class="section-title" id="header-submodules">Sub-modules</h2>
<dl>
<dt><code class="name"><a title="dadi.DFE" href="DFE/index.html">dadi.DFE</a></code></dt>
<dd>
<div class="desc"><p>Fitting Distributions of Fitness Effects to population genomic data, both single-population and joint.</p></div>
</dd>
<dt><code class="name"><a title="dadi.Demes" href="Demes/index.html">dadi.Demes</a></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt><code class="name"><a title="dadi.Demographics1D" href="Demographics1D.html">dadi.Demographics1D</a></code></dt>
<dd>
<div class="desc"><p>Single population demographic models.</p></div>
</dd>
<dt><code class="name"><a title="dadi.Demographics2D" href="Demographics2D.html">dadi.Demographics2D</a></code></dt>
<dd>
<div class="desc"><p>Two population demographic models.</p></div>
</dd>
<dt><code class="name"><a title="dadi.Godambe" href="Godambe.html">dadi.Godambe</a></code></dt>
<dd>
<div class="desc"><p>Parameter uncertainties and likelihood ratio tests using Godambe information.</p></div>
</dd>
<dt><code class="name"><a title="dadi.Inference" href="Inference.html">dadi.Inference</a></code></dt>
<dd>
<div class="desc"><p>Comparison and optimization of model spectra to data.</p></div>
</dd>
<dt><code class="name"><a title="dadi.Integration" href="Integration.html">dadi.Integration</a></code></dt>
<dd>
<div class="desc"><p>Functions for integrating population frequency spectra.</p></div>
</dd>
<dt><code class="name"><a title="dadi.Misc" href="Misc.html">dadi.Misc</a></code></dt>
<dd>
<div class="desc"><p>Miscellaneous utility functions. Including ms simulation.</p></div>
</dd>
<dt><code class="name"><a title="dadi.NLopt_mod" href="NLopt_mod.html">dadi.NLopt_mod</a></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt><code class="name"><a title="dadi.Numerics" href="Numerics.html">dadi.Numerics</a></code></dt>
<dd>
<div class="desc"><p>Numerically useful functions, including extrapolation and default grid.</p></div>
</dd>
<dt><code class="name"><a title="dadi.PhiManip" href="PhiManip.html">dadi.PhiManip</a></code></dt>
<dd>
<div class="desc"><p>Manipulating population frequency spectra phi. e.g. population splittings and
admixture</p></div>
</dd>
<dt><code class="name"><a title="dadi.Plotting" href="Plotting.html">dadi.Plotting</a></code></dt>
<dd>
<div class="desc"><p>Routines for Plotting comparisons between model and data …</p></div>
</dd>
<dt><code class="name"><a title="dadi.Spectrum_mod" href="Spectrum_mod.html">dadi.Spectrum_mod</a></code></dt>
<dd>
<div class="desc"><p>Contains Spectrum object, which represents frequency spectra.</p></div>
</dd>
<dt><code class="name"><a title="dadi.Triallele" href="Triallele/index.html">dadi.Triallele</a></code></dt>
<dd>
<div class="desc"><p>Modeling triallelic data.</p></div>
</dd>
<dt><code class="name"><a title="dadi.TwoLocus" href="TwoLocus/index.html">dadi.TwoLocus</a></code></dt>
<dd>
<div class="desc"><p>Fitting models to data summarized by pairs of loci, incorporating both allele frequencies and linkage disequilibrium</p></div>
</dd>
<dt><code class="name"><a title="dadi.integration_c" href="integration_c.html">dadi.integration_c</a></code></dt>
<dd>
<div class="desc"><p>This module 'integration_c' is auto-generated with f2py (version:1.21.2).</p>
<h2 id="functions">Functions</h2>
<p>phi = implicit_1Dx(phi,xx,nu,gamma,h,beta,dt,use_delj_trick)
…</p></div>
</dd>
<dt><code class="name"><a title="dadi.tridiag" href="tridiag.html">dadi.tridiag</a></code></dt>
<dd>
<div class="desc"><p>This module 'tridiag' is auto-generated with f2py (version:1.21.2).</p>
<h2 id="functions">Functions</h2>
<p>u = tridiag(a,b,c,r)
u = tridiag_fl(a,b,c,r)
.</p></div>
</dd>
</dl>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="dadi.RAM_to_pts"><code class="name flex">
<span>def <span class="ident">RAM_to_pts</span></span>(<span>RAM, P)</span>
</code></dt>
<dd>
<div class="desc"><p>Approximate maximum grid points given the number of populations and available RAM</p>
<p>pts: Grid points setting
P: Number of populations</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def RAM_to_pts(RAM, P):
    &#34;&#34;&#34;
    Approximate maximum grid points given the number of populations and available RAM

    pts: Grid points setting
    P: Number of populations
    &#34;&#34;&#34;
    return int((RAM*1024**3/(8*4))**(1./P))</code></pre>
</details>
</dd>
<dt id="dadi.citation"><code class="name flex">
<span>def <span class="ident">citation</span></span>(<span>)</span>
</code></dt>
<dd>
<div class="desc"><p>Print citation information for dadi codebase.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def citation():
    &#34;&#34;&#34;
    Print citation information for dadi codebase.
    &#34;&#34;&#34;
    print(&#34;&#34;&#34;
If you find dadi useful in your research, please cite:
RN Gutenkunst, RD Hernandez, SH Williamson, CD Bustamante &#34;Inferring the joint demographic history of multiple populations from multidimensional SNP data&#34; PLoS Genetics 5:e1000695 (2009)

If you find the Godambe Information Matrix methods useful, please cite:
AJ Coffman, P Hsieh, S Gravel, RN Gutenkunst &#34;Computationally efficient composite likelihood statistics for demographic inference&#34; Molecular Biology and Evolution 33:591 (2016)

If you find the DFE inference methods useful, please cite:
BY Kim, CD Huber, KE Lohmueller &#34;Inference of the Distribution of Selection Coefficients for New Nonsynonymous Mutations Using Large Samples&#34; Genetics 206:345 (2017)

If you find the triallelic methods useful, please cite:
AP Ragsdale, AJ Coffman, P Hsieh, TJ Struck, RN Gutenkunst &#34;Triallelic population genomics for inferring correlated fitness effects of same site nonsynonymous mutations&#34; Genetics 203:513 (2016)

If you find the two-locus methods useful, please cite:
AP Ragsdale, RN Gutenkunst &#34;Inferring demographic history using two-locus statistics&#34; Genetics 206:1037 (2017)
&#34;&#34;&#34;)</code></pre>
</details>
</dd>
<dt id="dadi.cuda_enabled"><code class="name flex">
<span>def <span class="ident">cuda_enabled</span></span>(<span>toggle=None)</span>
</code></dt>
<dd>
<div class="desc"><p>Enable or disable cuda execution</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def cuda_enabled(toggle=None):
    &#34;&#34;&#34;
    Enable or disable cuda execution
    &#34;&#34;&#34;
    if toggle is None:
        return Integration.cuda_enabled
    elif toggle == True:
        try:
            from . import cuda
            Integration.cuda_enabled = True
            return True
        except:
            print(&#34;Failed to import dadi.cuda&#34;)
            return False
    elif toggle == False:
        Integration.cuda_enabled = False
        return False
    else:
        raise ValueError(&#34;toggle must be True, False, or None&#34;)</code></pre>
</details>
</dd>
<dt id="dadi.pts_to_RAM"><code class="name flex">
<span>def <span class="ident">pts_to_RAM</span></span>(<span>pts, P)</span>
</code></dt>
<dd>
<div class="desc"><p>Approximate RAM usage for a given grid points and number of populations </p>
<p>pts: Grid points setting
P: Number of populations</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def pts_to_RAM(pts, P):
    &#34;&#34;&#34;
    Approximate RAM usage for a given grid points and number of populations 

    pts: Grid points setting
    P: Number of populations
    &#34;&#34;&#34;
    return 8*4*pts**P / 1024**3</code></pre>
</details>
</dd>
</dl>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3><a href="#header-submodules">Sub-modules</a></h3>
<ul>
<li><code><a title="dadi.DFE" href="DFE/index.html">dadi.DFE</a></code></li>
<li><code><a title="dadi.Demes" href="Demes/index.html">dadi.Demes</a></code></li>
<li><code><a title="dadi.Demographics1D" href="Demographics1D.html">dadi.Demographics1D</a></code></li>
<li><code><a title="dadi.Demographics2D" href="Demographics2D.html">dadi.Demographics2D</a></code></li>
<li><code><a title="dadi.Godambe" href="Godambe.html">dadi.Godambe</a></code></li>
<li><code><a title="dadi.Inference" href="Inference.html">dadi.Inference</a></code></li>
<li><code><a title="dadi.Integration" href="Integration.html">dadi.Integration</a></code></li>
<li><code><a title="dadi.Misc" href="Misc.html">dadi.Misc</a></code></li>
<li><code><a title="dadi.NLopt_mod" href="NLopt_mod.html">dadi.NLopt_mod</a></code></li>
<li><code><a title="dadi.Numerics" href="Numerics.html">dadi.Numerics</a></code></li>
<li><code><a title="dadi.PhiManip" href="PhiManip.html">dadi.PhiManip</a></code></li>
<li><code><a title="dadi.Plotting" href="Plotting.html">dadi.Plotting</a></code></li>
<li><code><a title="dadi.Spectrum_mod" href="Spectrum_mod.html">dadi.Spectrum_mod</a></code></li>
<li><code><a title="dadi.Triallele" href="Triallele/index.html">dadi.Triallele</a></code></li>
<li><code><a title="dadi.TwoLocus" href="TwoLocus/index.html">dadi.TwoLocus</a></code></li>
<li><code><a title="dadi.integration_c" href="integration_c.html">dadi.integration_c</a></code></li>
<li><code><a title="dadi.tridiag" href="tridiag.html">dadi.tridiag</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="dadi.RAM_to_pts" href="#dadi.RAM_to_pts">RAM_to_pts</a></code></li>
<li><code><a title="dadi.citation" href="#dadi.citation">citation</a></code></li>
<li><code><a title="dadi.cuda_enabled" href="#dadi.cuda_enabled">cuda_enabled</a></code></li>
<li><code><a title="dadi.pts_to_RAM" href="#dadi.pts_to_RAM">pts_to_RAM</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.8.1</a>.</p>
</footer>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
<script>hljs.initHighlightingOnLoad()</script>
</body>
</html>